src/main/java/org/testng/JarFileUtils.java

```diff
@@ -52,26 +59,59 @@ private boolean testngXmlExistsInJar(File jarFile, List<String> classes) throws
         boolean foundTestngXml = false;
         try (JarFile jf = new JarFile(jarFile)) {
             Enumeration<JarEntry> entries = jf.entries();
+            File file = createTempDir();
+            String suitePath = null;
             while (entries.hasMoreElements()) {
                 JarEntry je = entries.nextElement();
-                if (matchesXmlPathInJar(je)) {
-                    Collection<XmlSuite> parsedSuites = Parser.parse(jf.getInputStream(je), processor);
-                    for (XmlSuite suite : parsedSuites) {
-                        // If test names were specified, only run these test names
-                        XmlSuite suiteToAdd = suite;
-                        if (testNames != null) {
-                            suiteToAdd = XmlSuiteUtils.cloneIfContainsTestsWithNamesMatchingAny(suite, testNames);
-                        }
-                        suites.add(suiteToAdd);
+                if (je.getName().toLowerCase().endsWith(".xml")){
+                    String jeName = je.getName();
+                    InputStream  inputStream = jf.getInputStream(je);
+                    File copyFile = new File(file, jeName);
+                    Files.copyFile(inputStream, copyFile);
+                    if (matchesXmlPathInJar(je)) {
+                        suitePath = copyFile.toString();
                     }
-                    return true;
-                } else if (isJavaClass(je)) {
+                }else if (isJavaClass(je)) {
                     classes.add(constructClassName(je));
                 }
             }
+            if(Strings.isNullOrEmpty(suitePath)){
+                return foundTestngXml;
+            }
+            Collection<XmlSuite> parsedSuites = Parser.parse(suitePath,processor ); 
+            for (XmlSuite suite : parsedSuites) {
+                // If test names were specified, only run these test names
+                if (testNames != null) {
+                  XmlSuiteUtils xmlSuiteUitls = new XmlSuiteUtils();
+                  xmlSuiteUitls.cloneIfContainsTestsWithNamesMatchingAny(suite, testNames);
+                  List<String> missMatchedTestname = xmlSuiteUitls.getMissMatchedTestNames(testNames);
+                  if (CollectionUtils.hasElements(missMatchedTestname)) {
+                    throw new TestNGException("The test(s) <" + Arrays.toString(missMatchedTestname.toArray())+ "> cannot be found.");
+                  }
+                  suites.addAll(xmlSuiteUitls.getCloneSuite());
+                } else {
+                  suites.add(suite);
+                }
+                return true;
+              }
         }
         return foundTestngXml;
     }
+    
+    private static File createTempDir() {
+        File baseDir = new File(System.getProperty("java.io.tmpdir"));
```

Comment:
We can use Java7 helper methods: https://www.mkyong.com/java/how-to-create-temporary-file-in-java/

Reply:
Prefer my solution that refers from guava. How about I move it to org.testng.reporters.Files as util?

Reply:
Guava used it because they wanted to work on java 6 which is not our case. Why not taking the jdk api when it is available?

Reply:
Done.

Reply:
The two solutions have different purpose, I want to create a temporary folder, not a temporary file. 
`File.createTempFile` only able to create file.

Reply:
But yes, thanks for reminder. I would use `Files.createTempDirectory` instead.